docs/config.md
---
title: Configuration
description: Configuration for care.nvim
---

# Configuration

<details>
  <summary>Full Default Config</summary>

```lua
---@type care.config
config.defaults = {
    ui = {
        menu = {
            max_height = vim.o.pumheight ~= 0 and vim.o.pumheight or 10,
            border = "rounded",
            position = "auto",
            format_entry = function(entry, data)
                return require("care.presets").Default(entry, data)
            end,
            scrollbar = { enabled = true, character = "┃", offset = 0 },
            alignments = {},
        },
        docs_view = {
            max_height = 8,
            max_width = 80,
            border = "rounded",
            scrollbar = { enabled = true, character = "┃", offset = 0 },
            position = "auto",
            advanced_styling = false,
        },
        type_icons = {
            Class = "󰠱",
            Color = "󰏘",
            Constant = "󰏿",
            Constructor = "󰒓",
            Enum = "",
            EnumMember = "",
            Event = "",
            Field = "󰜢",
            File = "󰈚",
            Folder = "󰉋",
            Function = "󰆧",
            Interface = "󰙅",
            Keyword = "󰌋",
            Method = "󰆧",
            Module = "",
            Namespace = "󰌗",
            Operator = "󰆕",
            Property = "󰜢",
            Reference = "󰈇",
            Snippet = "",
            Struct = "󰙅",
            Text = "󰉿",
            TypeParameter = "󰊄",
            Unit = "󰑭",
            Value = "󰎠",
            Variable = "󰀫",
        },
        ghost_text = {
            enabled = true,
            position = "overlay",
        },
    },
    snippet_expansion = function(snippet_body)
        vim.snippet.expand(snippet_body)
    end,
    selection_behavior = "select",
    confirm_behavior = "insert",
    keyword_pattern = [[\%(-\?\d\+\%(\.\d\+\)\?\|\h\w*\%(-\w*\)*\)]],
    sources = {},
    preselect = true,
    sorting_direction = "top-down",
    completion_events = { "TextChangedI" },
    enabled = function()
        local enabled = true
        if vim.api.nvim_get_option_value("buftype", { buf = 0 }) == "prompt" then
            enabled = false
        end
        return enabled
    end,
    max_view_entries = 200,
    debug = false,
}
```

</details>

The config of care is used to configure the UI and care itself.
---
To configure care you call the setup function of the config module with the options you want
to override. Also see [configuration recipes](/configuration_recipes).
```lua
require("care").setup({
    ...
}
```
---
The configuration consists of two main parts. The UI Configuration and the
configuration of the completion behaviors of care.
# `care.config`

# Methods

## Snippet Expansion
`config.snippet_expansion?(body: string): nil`

With this field a function for expanding snippets is defined. By default this is the
builtin `vim.snippet.expand()`. You can also use a plugin like luasnip for this:
```lua
snippet_expansion = function(body)
    require("luasnip").lsp_expand(body)
end
```

## Enabled
`config.enabled?(): boolean`

This function can be used to disable care in certain contexts. By default this
disables care in prompts.
# Fields

## Ui
`config.ui? care.config.ui`

The [UI Configuration](#Ui-Configuration) is used to configure the whole UI of care.
One of the main goals of this is to be as extensible as possible. This is especially important
for the completion entries. Read more about that under
[Configuration of item display](/design/#configuration-of-item-display).

## Selection Behavior
`config.selection_behavior? "select"|"insert"`

With the selection behavior the user can determine what happens when selecting
an entry. This can either be `"select"` or `"insert"`. Selecting will just
select the entry and do nothing else. Insert will actually insert the text of
the entry (this is not necessarily the whole text).

## Confirm Behavior
`config.confirm_behavior? "insert"|"replace"`

This field controls the behavior when confirming an entry.

## Sources
`config.sources? table<string, care.config.source>`

See [care.config.source](/config/#source-configuration)

This field is used to configure the sources for care.nvim.
Use a table where the fields is the source name and the value is the configuration
```lua
sources = {
    nvim_lsp = {
        enabled = function()
            ...
        end,
        ...
    }
}
```

## Completion Events
`config.completion_events? string[]`

The `completion_events` table is used to set events for autocompletion. By default
it just contains `"TextChangedI"`. You can set it to an empty table (`{}`) to
disable autocompletion.

## Keyword Pattern
`config.keyword_pattern? string`

The keyword pattern is used to determine keywords. These are used to determine what
to use for filtering and what to remove if insert text is used.
It should essentially just describe the entries of a source.

## Preselect
`config.preselect? boolean`

Whether items should be preselected or not. Which items are preselected is determined
by the source.

## Sorting Direction
`config.sorting_direction? "top-down"|"away-from-cursor"`

How to sort the entries in the completion menu.
This can either be top to bottom or so the best match is always next to the cursor.

## Debug
`config.debug? boolean`

Whether debugging should be enabled or not. This will write a log to a `care.log` file
in the current directory.

## Max View Entries
`config.max_view_entries? integer`

The max entries to display in the menu. Note that this just affects
the entries displayed at a time. So there are still more entries on which you
can filter by continue typing. This limit just exists for performance reasons
because there are some sources which return up to multiple thousand
entries through which a user never will scroll.
The default value is 200 which will open the menu instantly in most cases so there
isn't much need for a lower value.
Values of up to about 1000 should also not cause performance issue.

# UI Configuration
This is used to configure the whole UI of care.
# `care.config.ui`

# Fields

## Menu
`config.ui.menu? care.config.ui.menu`

Configuration of the completion menu of care.nvim

## Docs View
`config.ui.docs_view? care.config.ui.docs`

This configuration allows you to configure the documentation view. It consists
of some basic window properties like the border and the maximum height of the
window. It also has a field to define the character used for the scrollbar.

## Type Icons
`config.ui.type_icons? care.config.ui.type_icons|string`

This is a table which defines the different icons.

## Ghost Text
`config.ui.ghost_text? care.config.ui.ghost_text`

Configuration of ghost text.

With this field the user can control how ghost text is displayed.
# `care.config.ui.ghost_text`

# Fields

## Enabled
`config.ui.ghost_text.enabled? boolean`

You can use the `enabled` field to determine whether the ghost text should be
enabled or not.

## Position
`config.ui.ghost_text.position? "inline"|"overlay"`

The `position` can either be `"inline"` or `"overlay"`. Inline
will add the text inline right where the cursor is. With the overlay position
the text will overlap with existing text after the cursor.

This configuration should allow you to completely adapt the completion menu to
your likings.
---
It includes some basic window properties like the border and the maximum height
of the window. It also has a field to define the character used for the
scrollbar. Set `scrollbar` to `nil` value to disable the scrollbar.
# `care.config.ui.menu`

# Methods

## Format Entry
`config.ui.menu.format_entry?(entry: care.entry, data: care.format_data): { [1]: string, [2]: string }[][]`

See [care.entry](/dev/entry)

Another field is `format_entry`. This is a function which recieves an entry of
the completion menu and determines how it's formatted. For that a table with
text-highlight chunks like `:h nvim_buf_set_extmarks()` is used. You can create
sections which are represented by tables and can have a different alignment
each. This is specified with another field which takes a table with the
alignment of each section.
For example you want to have the label of an entry in a red highlight and an
icon in an entry-kind specific color left aligned first and then the source of
the entry right aligned in blue. You could do that like this:
```lua
format_entry = function(entry)
    return {
        -- The first section with the two chunks for the label and the icon
        { { entry.label .. " ", "MyRedHlGroup" }, { entry.kind, "HighlightKind" .. entry.kind } }
        -- The second section for the source
        { { entry.source, "MyBlueHlGroup" } }
    }
end,
alignment = { "left", "right" }
```
Notice that there are multiple differences between having one table containing
the chunks for the label and kind and having them separately. The latter would
require another entry in the `alignment` table. It would also change the style
of the menu because the left sides of the icons would be aligned at the same
column and not be next to the labels. In the example there also was some spacing
added in between the two.
# Fields

## Max Height
`config.ui.menu.max_height? integer`

Maximum height of the menu

## Border
`config.ui.menu.border? string|string[]|string[][]`

The border of the completion menu

## Scrollbar
`config.ui.menu.scrollbar? care.config.scrollbar`

Configuration of the scrollbar

## Position
`config.ui.menu.position? "auto"|"below"|"above"`

If the menu should be displayed on above, below or automatically

## Alignments
`config.ui.menu.alignments? ("left"|"center"|"right")[]`

How the sections in the menu should be aligned

## Source configuration
Configuration for the sources of care.nvim
# `care.config.source`

# Methods

## Enabled
`config.source.enabled? boolean|fun():boolean`

Whether the source is enabled (default true)

## Filter
`config.source.filter?(entry: care.entry): boolean`

See [care.entry](/dev/entry)

Filter function for entries by the source
# Fields

## Max Entries
`config.source.max_entries? integer`

The maximum amount? of entries which can be displayed by this source

## Priority
`config.source.priority? integer`

The priority of this source. Is more important than matching score

Configuration of the completion menu of care.nvim
# `care.config.ui.docs`

# Fields

## Max Height
`config.ui.docs.max_height? integer`

Maximum height of the documentation view

## Max Width
`config.ui.docs.max_width? integer`

Maximum width of the documentation view

## Border
`config.ui.docs.border? string|string[]|string[][]`

The border of the documentation view

## Scrollbar
`config.ui.docs.scrollbar? care.config.scrollbar`

Configuration of the scrollbar

## Position
`config.ui.docs.position? "auto"|"left"|"right"`

Position of docs view.
Auto will prefer right if there is enough space

## Advanced Styling
`config.ui.docs.advanced_styling? boolean`

Use advanced styling for the documentation.
This will use treesitter to make the menu prettier which comes with quite a big
performance cost though.

Additional data passed to format function to allow more advanced formatting
# `care.format_data`

# Fields

## Index
`format_data.index integer`

Index of the entry in the completion menu

## Deprecated
`format_data.deprecated boolean`

Whether the item is marked as deprecated by the source or not

## Source Name
`format_data.source_name string`

The name of the source from which the entry was completed

## Source Display Name
`format_data.source_display_name string`

The display name of the source from which the entry was completed which
is more detailed than the normal name


# `care.config.scrollbar`

# Fields

## Character
`config.scrollbar.character? string`

The character used for drawing the scrollbar

## Enabled
`config.scrollbar.enabled? boolean`

Whether the scrollbar is enabled or not

## Offset
`config.scrollbar.offset? integer`

Offset of the scrollbar. 0 is at the border of the window

docs/configuration_recipes.md
---
title: Configuration Recipes
description: Some common configuration recipes for care.nvim
---

# Configuration recipes

Here are some useful configurations for care.

## Common fields

This section lists some commonly uses fields in which users might be interested.

### Source name

You can access the name of the source form which the entry was completed with
the `source_name` or the `source_display_name` field in the additional data
passed to the `format_entry` function. The display name will often be
significantly longer so it's recommended to use `source_name`.

## Labels and shortcuts

This configuration will add labels to your items and allow to select them with
`ctrl+label`.

```lua
local labels = {"1", "2", "3", "4", "5", "6", "7", "8", "9"}

-- Add this to formatting

{ {
    " " .. require("care.presets.utils").label_entries(labels)(entry, data) .. " ",
    "Comment",
}, },

-- Keymappings
for i, label in ipairs(labels) do
    vim.keymap.set("i", "<c-"..label..">", function()
        require("care").api.select_visible(i)
        -- If you also want to confirm the entry
        require("care").api.confirm()
    end)
end

```

## Access Documentation

You can use these function to access documentation. This e.g. allows to copy it
to somewhere for reference or to click links in it easily.

They have to be called while the documentation window of care is actually open.

### Open in split

This function opens the documentation in a split window.

```lua
local documentation = require("care").api.get_documentation()
if #documentation == 0 then
    return
end
local old_win = vim.api.nvim_get_current_win()
vim.cmd.wincmd("s")
local buf = vim.api.nvim_create_buf(false, true)
vim.bo[buf].ft = "markdown"
vim.api.nvim_buf_set_lines(buf, 0, -1, false, documentation)
vim.api.nvim_win_set_buf(0, buf)
vim.api.nvim_set_current_win(old_win)
```

### Copy to clipboard

This function copies the documentation to the clipboard (or any other register).

```lua
local documentation = require("care").api.get_documentation()
if #documentation == 0 then
    return
end
vim.fn.setreg("+", table.concat(documentation, "\n"))
```

## Integrations for icons

There are various integrations with different plugins for icons in care.nvim

### Mini.icons

Install mini.icons as described in
[the readme](https://github.com/echasnovski/mini.icons/tree/main). (Don't forget
to call `require"mini.icons".setup()`!).

Then use the following configuration for care.

```lua
require("care").setup({
    ui = {
        type_icons = "mini.icons"
    }
})
```

Mini.icons also allows you to use ascii icons by using `{ style = "ascii" }` in
the setup.

### Lspkind.nvim

Install lspkind.nvim as described in
[the readme](https://github.com/onsails/lspkind.nvim/tree/master).

Then use the following configuration for care.

```lua
require("care").setup({
    ui = {
        type_icons = "lspkind"
    }
})
```

## Manual completion like builtin neovim

You can use filters to only complete certain sources. This can be used to create
a behavior like builtin neovim.

There are always certain sources required for the mappings.

You likely also want to set [completion_events](/config#completion-events) to an
empty table `{}` to disable autocompletion if you want this behavior.

### LSP Omnifunc

You can create something similar to the behavior of setting omnifunc to
`vim.lsp.omnifunc()` like this:

```lua
-- Source: none, is builtin
vim.keymap.set("i", "<c-x><c-o>", function()
    require("care").api.complete(function(name)
        return name == "lsp"
    end)
end)
```

### Paths

```lua
-- Source: "hrsh7th/cmp-path" (requires "max397574/care-cmp")
-- Limitations: In comparison to builtin completion the pattern to find filenames is different
-- So you have to use e.g. `./in` to complete `init.lua` instead of just `in` like builting completion

vim.keymap.set("i", "<c-x><c-f>", function()
    require("care").api.complete(function(name)
        return name == "cmp_path"
    end)
end)
```

### Buffer keywords

```lua
-- Source: "hrsh7th/cmp-buffer" (requires "max397574/care-cmp")
-- Limitations: Searches in whole buffers, forwards and backwards cursor

-- same for "<c-x><c-n>", "<c-x><c-p>"
vim.keymap.set("i", "<c-x><c-i>", function()
    require("care").api.complete(function(name)
        return name == "cmp_buffer"
    end)
end)
```

### Buffer lines

```lua
-- Source: "amarakon/nvim-cmp-buffer-lines" (requires "max397574/care-cmp")
-- Limitations: Not tested enough to know

vim.keymap.set("i", "<c-x><c-l>", function()
    require("care").api.complete(function(name)
        return name == "cmp_buffer-lines"
    end)
end)
```

### Spelling

```lua
-- Source: "f3fora/cmp-spell" (requires "max397574/care-cmp")
-- Limitations: Not tested enough to know

vim.keymap.set("i", "<c-x><c-s>", function()
    require("care").api.complete(function(name)
        return name == "cmp_spell"
    end)
end)
```

## Reverse keybindings for reversed menu

When using `"away-from-cursor"` as `sorting_direction` in the configuration the
menu can be reversed. The `select_{next, prev}` mappings will still go in the
same direction though. If you always want to use `select_next` to select the
entry with a lower match than current one you can use mappings like this:

```lua
vim.keymap.set("i", "<c-n>", function()
    if require("care").api.is_reversed() then
        require("care").api.select_prev()
    else
        require("care").api.select_next()
    end
end)

vim.keymap.set("i", "<c-p>", function()
    if require("care").api.is_reversed() then
        require("care").api.select_next()
    else
        require("care").api.select_prev()
    end
end)
```

## Use with nvim-autopairs

```lua
"windwp/nvim-autopairs",
lazy = false,
config = function()
    local npairs = require("nvim-autopairs")
    npairs.setup({ map_cr = false })
    vim.keymap.set("i", "<cr>", function()
        if require("care").api.is_open() then
            require("care").api.confirm()
        else
            vim.fn.feedkeys(require("nvim-autopairs").autopairs_cr(), "in")
        end
    end)
end,
```

## Common Configurations

### Luasnip and friendly-snippets

-   Optional: Install
    [friendly-snippets](https://github.com/rafamadriz/friendly-snippets)
-   Install [LuaSnip](https://github.com/L3MON4D3/LuaSnip)
-   Install [care-cmp](https://github.com/max397574/care-cmp)
-   Install [cmp_luasnip](https://github.com/saadparwaiz1/cmp_luasnip)

<details>
<summary>Lazy.nvim Example</summary>

```lua
{
    "max397574/care.nvim",
    dependencies = {
        "max397574/care-cmp",
        "saadparwaiz1/cmp_luasnip",
        "rafamadriz/friendly-snippets",
        "L3MON4D3/LuaSnip",
    },
    config = function()
        ...
    end
}
```

</details>

Then you can use the following configuration to use luasnip for snippet
completion and to load the friendly-snippets snippets.

```lua
config = function()
    require("luasnip.loaders.from_vscode").lazy_load()
    require("care").setup({
        snippet_expansion = function(body)
            require("luasnip").lsp_expand(body)
        end,
        ...
    })
end
```


CHANGELOG.md
# Changelog

## [1.0.0](https://github.com/max397574/care.nvim/compare/v0.1.0...v1.0.0) (2024-12-11)


### ⚠ BREAKING CHANGES

* **presets:** rename some presets and extend a lot
* better scrollbar config
* rename lsp source to just lsp
* rename setup to initialization and use setup for configuration
* only use config module for configuration

### Features

* adapt menu to changes in window class ([c9f4c85](https://github.com/max397574/care.nvim/commit/c9f4c85b467f1a4fd13ae33240dc3115bf1cbc89))
* adapt to code style ([f2a3930](https://github.com/max397574/care.nvim/commit/f2a393004282401a8248025a9284408b97b93752))
* add api to check if menu is reversed ([b5199b1](https://github.com/max397574/care.nvim/commit/b5199b1708145de6a6c2537c41e33a280e528490))
* add away from cursor sorting ([8cfea18](https://github.com/max397574/care.nvim/commit/8cfea184fef21c6f1fa726b6ce22e7b47f8e53d4))
* add code style document ([8d250ae](https://github.com/max397574/care.nvim/commit/8d250ae3c2158a14cf9e4e9b0698d2fd97c52022))
* add colored block to completion menu ([aea434f](https://github.com/max397574/care.nvim/commit/aea434f0e14bf13fee1cd3af752032f920782e63))
* add config option to limit displayed entries ([d486485](https://github.com/max397574/care.nvim/commit/d4864853085b2e8273df4de27aae22cb762cf5b0))
* add deprecated as data to format ([964cc22](https://github.com/max397574/care.nvim/commit/964cc22648dd5e1f61196ac26df2402b366adcfb))
* add editorconfig ([44a33da](https://github.com/max397574/care.nvim/commit/44a33da03a169507a55760b3137db74c6b643c3a))
* add execute to sources ([4f1a828](https://github.com/max397574/care.nvim/commit/4f1a8284e7c1686549beb00aac0a9d7001654bc3))
* add initial healthcheck ([a201df9](https://github.com/max397574/care.nvim/commit/a201df94ae87b9f5720186ed1bb12510d3d5b127))
* add logo ([d851715](https://github.com/max397574/care.nvim/commit/d8517153fd839f74ff9c5f130ba53c5f3c1b1ef5))
* add lsp source to main repo ([74e5fd3](https://github.com/max397574/care.nvim/commit/74e5fd3fd339841679ced2a592a0d54700638b35))
* add luasnip configuration recipe ([61f564c](https://github.com/max397574/care.nvim/commit/61f564c71ac3642f4297428e3bdb03af6b1ebd4a))
* add minimal bug report template ([b138fd4](https://github.com/max397574/care.nvim/commit/b138fd4d011df831b1710d42796c9ddb8b7b7f95))
* add new api function `get_index` ([5c2bbdb](https://github.com/max397574/care.nvim/commit/5c2bbdbebe5565f404921b0416c25ea2a7c78193))
* add path source ([d3b0642](https://github.com/max397574/care.nvim/commit/d3b0642b21efd48c4159f4dc969a525944608611))
* add preselect support ([f949bb3](https://github.com/max397574/care.nvim/commit/f949bb3563fd98f5d4b7c8a64780cf6c4edba95b))
* add preset utils and recipes ([6cf8c1a](https://github.com/max397574/care.nvim/commit/6cf8c1a57165636331eab6dbfaacc769f602575a))
* add test for new bug ([e124078](https://github.com/max397574/care.nvim/commit/e124078b6c242b1e7b623bfb6deee1c8e220ab99))
* add type annotation to config setup ([5663b3a](https://github.com/max397574/care.nvim/commit/5663b3a472875e3b6b3a43a5121652f452e7bc2a))
* add type annotations to care.setup ([e32beab](https://github.com/max397574/care.nvim/commit/e32beabfe734dbe44818355c2f4a64a8ae8fe7db))
* added filter to core ([98e7c9c](https://github.com/max397574/care.nvim/commit/98e7c9c17b569a902285a12e8a3a5b4e336b400a))
* allow accessing documentation ([e70df3b](https://github.com/max397574/care.nvim/commit/e70df3b49558848a5a7547466a2af2f84dd46b77))
* allow filtering sources for completion ([aaab984](https://github.com/max397574/care.nvim/commit/aaab984dcf6b849c5834712d7d5a2619d7ef4c86))
* allow insert and replace edit range ([cb5577c](https://github.com/max397574/care.nvim/commit/cb5577c3751fa86e4346441afa192fc76f2d6b22))
* allow multiline ghost text ([ecb1b51](https://github.com/max397574/care.nvim/commit/ecb1b5179db241db78a019be2b5242acd4eacb0b))
* **api:** add function to select visible entry ([15a40a0](https://github.com/max397574/care.nvim/commit/15a40a066e981bf9e2fe7e65f1d4bb69e942e9fb))
* **api:** add function to set index ([90d96b4](https://github.com/max397574/care.nvim/commit/90d96b4b7480e8d3ae67fcfa98d6b134c7d6d2fb))
* autocomplete with trigger chars after confirm ([5df526c](https://github.com/max397574/care.nvim/commit/5df526c148e630dd550956c18463445a98e81f82))
* avoid completing and filtering at the same time ([941e8d9](https://github.com/max397574/care.nvim/commit/941e8d9cfb5d505f3c2d63dbe44d3f998c8b4a66))
* avoid duplicate draw ([0fc5ea9](https://github.com/max397574/care.nvim/commit/0fc5ea9568e45b67b3d78141d9170a17804ae723))
* better calculations for window width and height ([0bea5e4](https://github.com/max397574/care.nvim/commit/0bea5e46ed61fee6c1b8663666cbc879b918d435))
* better confirming and debugging ([b1a268f](https://github.com/max397574/care.nvim/commit/b1a268fc325e53a6619ff056f9a7b291411e5e1b))
* better docs for presets ([b452039](https://github.com/max397574/care.nvim/commit/b452039f3ad3750ef432f58dd9eedb074715537e))
* better logging of entry ([3b34a9f](https://github.com/max397574/care.nvim/commit/3b34a9f995039722966a293769c237a51d3a58e7))
* better scrollbar config ([77ce911](https://github.com/max397574/care.nvim/commit/77ce911ca4cf0de9cfc9154903fc30fda6bddea0))
* better width and position calculation for docs view ([c2dfc9a](https://github.com/max397574/care.nvim/commit/c2dfc9abeb40b1adfb7ada68faf3740cc00d39bd))
* check if source.execute is nil before executing ([91d1734](https://github.com/max397574/care.nvim/commit/91d1734150b5a0a76f5e8c92cb6a3bdd228b95f8))
* **config:** better icons ([fe02e6c](https://github.com/max397574/care.nvim/commit/fe02e6cf7c778d72e08da483a5f1ca771239dab1))
* **config:** default to pumheight as max height of menu ([dd11530](https://github.com/max397574/care.nvim/commit/dd11530859b34cf54574c93bb0ecbbb13f4341f0))
* **config:** mark deprecated items specially ([3cd8202](https://github.com/max397574/care.nvim/commit/3cd8202cd4557832a64242e00898b88e58ed3dbb))

lua/care/types/config.lua
--- The config of care is used to configure the UI and care itself.
---
--- To configure care you call the setup function of the config module with the options you want
--- to override. Also see [configuration recipes](/configuration_recipes).
--- ```lua
--- require("care").setup({
---     ...
--- }
--- ```
---
--- The configuration consists of two main parts. The UI Configuration and the
--- configuration of the completion behaviors of care.
---@class care.config
--- The [UI Configuration](#Ui-Configuration) is used to configure the whole UI of care.
--- One of the main goals of this is to be as extensible as possible. This is especially important
--- for the completion entries. Read more about that under
--- [Configuration of item display](/design/#configuration-of-item-display).
---@field ui? care.config.ui
--- With this field a function for expanding snippets is defined. By default this is the
--- builtin `vim.snippet.expand()`. You can also use a plugin like luasnip for this:
---
--- ```lua
--- snippet_expansion = function(body)
---     require("luasnip").lsp_expand(body)
--- end
--- ```
---@field snippet_expansion? fun(body: string): nil
--- With the selection behavior the user can determine what happens when selecting
--- an entry. This can either be `"select"` or `"insert"`. Selecting will just
--- select the entry and do nothing else. Insert will actually insert the text of
--- the entry (this is not necessarily the whole text).
---@field selection_behavior? "select"|"insert"
--- This field controls the behavior when confirming an entry.
---@field confirm_behavior? "insert"|"replace"
--- This field is used to configure the sources for care.nvim.
--- Use a table where the fields is the source name and the value is the configuration
--- ```lua
--- sources = {
---     nvim_lsp = {
---         enabled = function()
---             ...
---         end,
---         ...
---     }
--- }
--- ```
---@field sources? table<string, care.config.source>
--- The `completion_events` table is used to set events for autocompletion. By default
--- it just contains `"TextChangedI"`. You can set it to an empty table (`{}`) to
--- disable autocompletion.
---@field completion_events? string[]
--- The keyword pattern is used to determine keywords. These are used to determine what
--- to use for filtering and what to remove if insert text is used.
--- It should essentially just describe the entries of a source.
---@field keyword_pattern? string
--- This function can be used to disable care in certain contexts. By default this
--- disables care in prompts.
---@field enabled? fun(): boolean
--- Whether items should be preselected or not. Which items are preselected is determined
--- by the source.
---@field preselect? boolean
--- How to sort the entries in the completion menu.
--- This can either be top to bottom or so the best match is always next to the cursor.
---@field sorting_direction? "top-down"|"away-from-cursor"
--- Whether debugging should be enabled or not. This will write a log to a `care.log` file
--- in the current directory.
---@field debug? boolean
--- The max entries to display in the menu. Note that this just affects
--- the entries displayed at a time. So there are still more entries on which you
--- can filter by continue typing. This limit just exists for performance reasons
--- because there are some sources which return up to multiple thousand
--- entries through which a user never will scroll.
--- The default value is 200 which will open the menu instantly in most cases so there
--- isn't much need for a lower value.
--- Values of up to about 1000 should also not cause performance issue.
---@field max_view_entries? integer

--- # UI Configuration
--- This is used to configure the whole UI of care.
---@class care.config.ui
--- Configuration of the completion menu of care.nvim
---@field menu? care.config.ui.menu
--- This configuration allows you to configure the documentation view. It consists
--- of some basic window properties like the border and the maximum height of the
--- window. It also has a field to define the character used for the scrollbar.
---@field docs_view? care.config.ui.docs
--- This is a table which defines the different icons.
---@field type_icons? care.config.ui.type_icons|string
--- Configuration of ghost text.
---@field ghost_text? care.config.ui.ghost_text

--- With this field the user can control how ghost text is displayed.
---@class care.config.ui.ghost_text
--- You can use the `enabled` field to determine whether the ghost text should be
--- enabled or not.
---@field enabled? boolean
--- The `position` can either be `"inline"` or `"overlay"`. Inline
--- will add the text inline right where the cursor is. With the overlay position
--- the text will overlap with existing text after the cursor.
---@field position? "inline"|"overlay"

--- This configuration should allow you to completely adapt the completion menu to
--- your likings.
---
--- It includes some basic window properties like the border and the maximum height
--- of the window. It also has a field to define the character used for the
--- scrollbar. Set `scrollbar` to `nil` value to disable the scrollbar.
---@class care.config.ui.menu
--- Maximum height of the menu
---@field max_height? integer
--- The border of the completion menu
---@field border? string|string[]|string[][]
--- Configuration of the scrollbar
---@field scrollbar? care.config.scrollbar
--- If the menu should be displayed on above, below or automatically
---@field position? "auto"|"below"|"above"
--- Another field is `format_entry`. This is a function which recieves an entry of
--- the completion menu and determines how it's formatted. For that a table with
--- text-highlight chunks like `:h nvim_buf_set_extmarks()` is used. You can create
--- sections which are represented by tables and can have a different alignment
--- each. This is specified with another field which takes a table with the
--- alignment of each section.
---
--- For example you want to have the label of an entry in a red highlight and an
--- icon in an entry-kind specific color left aligned first and then the source of
--- the entry right aligned in blue. You could do that like this:
---
--- ```lua
--- format_entry = function(entry)
---     return {
---         -- The first section with the two chunks for the label and the icon
---         { { entry.label .. " ", "MyRedHlGroup" }, { entry.kind, "HighlightKind" .. entry.kind } }
---         -- The second section for the source
---         { { entry.source, "MyBlueHlGroup" } }
---     }
--- end,
--- alignment = { "left", "right" }
--- ```
---
--- Notice that there are multiple differences between having one table containing
--- the chunks for the label and kind and having them separately. The latter would
--- require another entry in the `alignment` table. It would also change the style
--- of the menu because the left sides of the icons would be aligned at the same
--- column and not be next to the labels. In the example there also was some spacing
--- added in between the two.
---@field format_entry? fun(entry: care.entry, data: care.format_data): { [1]: string, [2]: string }[][]
--- How the sections in the menu should be aligned
---@field alignments? ("left"|"center"|"right")[]

--- ## Source configuration
--- Configuration for the sources of care.nvim
---@class care.config.source
--- Whether the source is enabled (default true)
---@field enabled? boolean|fun():boolean
--- The maximum amount? of entries which can be displayed by this source
---@field max_entries? integer
--- The priority of this source. Is more important than matching score
---@field priority? integer
--- Filter function for entries by the source
---@field filter? fun(entry: care.entry): boolean

--- Configuration of the completion menu of care.nvim
---@class care.config.ui.docs
--- Maximum height of the documentation view
---@field max_height? integer
--- Maximum width of the documentation view
---@field max_width? integer
--- The border of the documentation view
---@field border? string|string[]|string[][]
--- Configuration of the scrollbar
---@field scrollbar? care.config.scrollbar
--- Position of docs view.
--- Auto will prefer right if there is enough space
---@field position? "auto"|"left"|"right"
--- Use advanced styling for the documentation.
--- This will use treesitter to make the menu prettier which comes with quite a big
--- performance cost though.
---@field advanced_styling? boolean

--- Additional data passed to format function to allow more advanced formatting
---@class care.format_data
--- Index of the entry in the completion menu
---@field index integer
--- Whether the item is marked as deprecated by the source or not
---@field deprecated boolean
--- The name of the source from which the entry was completed
---@field source_name string
--- The display name of the source from which the entry was completed which
--- is more detailed than the normal name
---@field source_display_name string

---@class care.config.scrollbar
--- The character used for drawing the scrollbar
---@field character? string
--- Whether the scrollbar is enabled or not
---@field enabled? boolean
--- Offset of the scrollbar. 0 is at the border of the window
---@field offset? integer
